File: apps/web/src/api/history.ts
import { ApiError, encodePath } from './client.ts';
import type { Note } from './client.ts';

/**
 * Typed client for the vault's version history.
 *
 * The history is an ordinary git repository in the vault folder, kept by the
 * server. Two shapes of question, and both matter: "what did this note say
 * before?" — the one people actually ask — and "put the whole vault back to
 * Tuesday", which is the recovery case.
 *
 * Nothing here rewrites anything. A rollback is a new version whose contents
 * are the old one's, so what it replaced stays reachable and rolling back the
 * rollback is the same action again. That is what makes it safe to try.
 */

/** What kind of moment a version records. "" for a commit made outside
 * Thought Mesh — someone's own `git commit` belongs in this history too. */
export type HistoryKind =
  | 'edit'
  | 'local'
  | 'sync'
  | 'conflict'
  | 'rollback'
  | 'checkpoint'
  | 'restore'
  | '';

/** One version of the vault. */
export interface HistoryCommit {
  ref: string;
  short: string;
  subject: string;
  /** Whatever was typed when this moment was marked; "" for most entries. */
  body: string;
  kind: HistoryKind;
  author: string;
  at_ms: number;
}

export interface HistoryState {
  /**
   * 0 when this server keeps no history — git isn't installed, or the
   * deployment turned it off. The routes still answer, so the client can tell
   * that apart from a server too old to have the feature.
   */
  available: 0 | 1;
  commits: HistoryCommit[];
}

async function request<T>(method: string, path: string, body?: unknown): Promise<T> {
  const init: RequestInit = { method };
  if (body !== undefined) {
    init.headers = { 'Content-Type': 'application/json' };
    init.body = JSON.stringify(body);
  }
  const res = await fetch(`/api${path}`, init);
  const text = await res.text();
  const data = text ? (JSON.parse(text) as unknown) : undefined;
  if (!res.ok) {
    const message =
      data && typeof data === 'object' && 'error' in data
        ? String((data as { error: unknown }).error)
        : `request failed (${res.status})`;
    throw new ApiError(res.status, message);
  }
  return data as T;
}

/** The vault's recent history, newest first. */
export function fetchHistory(limit?: number): Promise<HistoryState> {
  const query = limit ? `?limit=${limit}` : '';
  return request('GET', `/history${query}`);
}

/** The versions that touched one note — a note's own history. */
export function fetchNoteHistory(path: string, limit?: number): Promise<HistoryState> {
  const query = limit ? `?limit=${limit}` : '';
  return request('GET', `/history/notes/${encodePath(path)}${query}`);
}

/** One note's content as of a version. Reading only: the note is untouched. */
export function fetchNoteVersion(
  path: string,
  ref: string,
): Promise<{ path: string; ref: string; content: string }> {
  return request('GET', `/history/version/${encodePath(path)}?ref=${encodeURIComponent(ref)}`);
}

/**
 * Put one note back to an older version, leaving the rest of the vault alone.
 *
 * It goes through the ordinary note save on the server, so everything a save
 * does still happens — links are re-indexed, and the next cloud sync carries it
 * like any other edit. What was in between is not erased; it stays in the log.
 */
export function restoreNoteVersion(path: string, ref: string): Promise<Note> {
  return request('POST', '/history/restore', { path, ref });
}

/** Mark this moment, with an optional message. Recorded even when nothing has
 * changed — the message is the point. */
export function checkpointHistory(message: string): Promise<HistoryState> {
  return request('POST', '/history/checkpoint', { message });
}

/** Put the whole vault back to a version, as a new version. */
export function rollbackHistory(ref: string): Promise<HistoryState> {
  return request('POST', '/history/rollback', { ref });
}

/** How each kind of version is labelled. Absent (a hand-made commit) reads as
 * what it is: something that happened outside the app. */
export const HISTORY_KIND_LABELS: Record<HistoryKind, string> = {
  edit: 'Edited',
  local: 'Before a sync',
  sync: 'Synced',
  conflict: 'Conflict settled',
  rollback: 'Rolled back',
  checkpoint: 'Checkpoint',
  restore: 'Restored',
  '': 'Outside the app',
};

/**
 * The part of a version's subject worth showing next to its timestamp.
 *
 * Commit subjects carry the time on purpose — they are read with `git log`,
 * where there is no column of dates beside them. In this list there is, so the
 * stamp is stripped, and a subject that then says nothing the kind label
 * doesn't ("Checkpoint") is dropped entirely rather than repeated.
 */
export function historyDetail(commit: HistoryCommit): string {
  const withoutStamp = commit.subject
    .replace(/,?\s+at\s+\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}\s*$/, '')
    .trim();
  const label = HISTORY_KIND_LABELS[commit.kind];
  if (withoutStamp === '' || withoutStamp.toLowerCase() === label.toLowerCase()) {
    return '';
  }
  return withoutStamp;
}

File: server/internal/api/history.go
package api

import (
	"errors"
	"net/http"
	"strconv"
	"strings"

	"github.com/chinmay28/thought-mesh/server/internal/history"
	"github.com/chinmay28/thought-mesh/server/internal/vault"
)

// Version history endpoints, over the vault's git repository.
//
// Two shapes of question, and both matter: "what did this note say before?"
// (per-note, the one people actually ask) and "put the whole vault back to
// Tuesday" (the recovery case). The first is served by a note's own log plus
// reading it at a revision; the second by the vault log plus a rollback.
//
// Nothing here rewrites history. A rollback is a new commit whose tree is the
// old one, so the state being replaced stays reachable and rolling back the
// rollback is the same operation again — which is what makes it safe to try.
//
// Where git isn't installed the routes are still registered and answer
// honestly (`available: 0`, and 400 on the write paths) rather than 404, so
// the client can tell "this server has no history" from "this server is too
// old to have the feature".

// historyStatus is the wire shape of "is there history here at all".
type historyStatus struct {
	Available int `json:"available"`
	// Commits is the recent log, newest first. Empty when unavailable.
	Commits []history.Commit `json:"commits"`
}

// defaultHistoryLimit is how much log a request gets without asking. A screen's
// worth; `?limit=` raises it up to the package's own cap.
const defaultHistoryLimit = 50

func historyLimit(r *http.Request) int {
	raw := r.URL.Query().Get("limit")
	if raw == "" {
		return defaultHistoryLimit
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n <= 0 {
		return defaultHistoryLimit
	}
	return n
}

// handleHistoryErr maps this package's errors on top of the shared ones: a bad
// ref or path is a 400, an unavailable history is a 400 that says why, and a
// commit or path git doesn't know is a 404 rather than an internal error —
// asking for a note at a revision that predates it is an ordinary thing to do.
func handleHistoryErr(w http.ResponseWriter, err error) {
	var ve *history.ValidationError
	var ce *history.CommandError
	switch {
	case errors.Is(err, history.ErrUnavailable):
		writeErr(w, http.StatusBadRequest, err.Error())
	case errors.As(err, &ve):
		writeErr(w, http.StatusBadRequest, ve.Error())
	case errors.As(err, &ce):
		if isUnknownToGit(ce.Stderr) {
			writeErr(w, http.StatusNotFound, "no such version of that note")
			return
		}
		handleErr(w, err)
	default:
		handleErr(w, err)
	}
}

// isUnknownToGit recognizes "that object doesn't exist" among git's failures.
func isUnknownToGit(stderr string) bool {
	lower := strings.ToLower(stderr)
	for _, phrase := range []string{
		"exists on disk, but not in",
		"unknown revision",
		"does not exist",
		"invalid object name",
		"bad revision",
		"no such path",
	} {
		if strings.Contains(lower, phrase) {
			return true
		}
	}
	return false
}

// listHistory returns the vault's recent history.
func (s *server) listHistory(w http.ResponseWriter, r *http.Request) {
	body := historyStatus{Commits: []history.Commit{}}
	if !s.history.Available() {
		writeJSON(w, http.StatusOK, body)
		return
	}
	commits, err := s.history.Log(historyLimit(r))
	if err != nil {
		handleHistoryErr(w, err)
		return
	}
	body.Available, body.Commits = 1, commits
	writeJSON(w, http.StatusOK, body)
}

// noteHistory returns the commits that touched one note — the question people
// actually ask of a history.
func (s *server) noteHistory(w http.ResponseWriter, r *http.Request) {
	body := historyStatus{Commits: []history.Commit{}}
	if !s.history.Available() {
		writeJSON(w, http.StatusOK, body)
		return
	}
	commits, err := s.history.FileLog(r.PathValue("path"), historyLimit(r))
	if err != nil {
		handleHistoryErr(w, err)
		return
	}
	body.Available, body.Commits = 1, commits
	writeJSON(w, http.StatusOK, body)
}

// showNoteVersion returns a note's content as of one commit. Reading only —
// the working copy is untouched, so comparing costs nothing.
func (s *server) showNoteVersion(w http.ResponseWriter, r *http.Request) {
	ref := r.URL.Query().Get("ref")
	content, err := s.history.Show(ref, r.PathValue("path"))
	if err != nil {
		handleHistoryErr(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"path":    r.PathValue("path"),
		"ref":     ref,
		"content": string(content),
	})
}

// restoreNoteVersion puts one note back to an older version, leaving the rest
// of the vault alone.
//
// It goes through the ordinary note save rather than through git, so everything
// a save does still happens — the link index sees it, and the next cloud sync
// pushes it like any other edit. The old version becomes the current one; it
// does not erase what was in between, which is still in the log.
func (s *server) restoreNoteVersion(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Path string `json:"path"`
		Ref  string `json:"ref"`
	}
	if err := decodeBody(r, &body); err != nil {
		handleErr(w, err)
		return
	}
	content, err := s.history.Show(body.Ref, body.Path)
	if err != nil {
		handleHistoryErr(w, err)
		return
	}
	info, err := s.v.Write(body.Path, string(content))
	if err != nil {
		handleErr(w, err)
		return
	}
	short := body.Ref
	if len(short) > 7 {
		short = short[:7]
	}
	if _, err := s.history.Commit(
		"Restore "+info.Path+" to "+short+", at "+s.history.Stamp(), "", history.KindRollback,
	); err != nil {
		handleHistoryErr(w, err)
		return
	}
	note, err := s.fullNote(info.Path)
	if err != nil {
		handleErr(w, err)
		return
	}
	writeJSON(w, http.StatusOK, note)
}

// checkpointHistory marks a moment deliberately, with an optional message.
//
// It commits even when nothing has changed: the message *is* the point, and
// dropping it because the vault happened to be clean would be the surprising
// outcome.
func (s *server) checkpointHistory(w http.ResponseWriter, r *http.Request) {
	if !s.history.Available() {
		writeErr(w, http.StatusBadRequest, history.ErrUnavailable.Error())
		return
	}
	var body struct {
		Message string `json:"message"`
	}
	if err := decodeBody(r, &body); err != nil {
		handleErr(w, err)
		return
	}
	message := strings.TrimSpace(body.Message)
	subject := "Checkpoint at " + s.history.Stamp()
	if _, err := s.history.CommitAlways(subject, message, history.KindCheckpoint); err != nil {
		handleHistoryErr(w, err)
		return
	}
	s.writeHistory(w, r)
}

// rollbackHistory puts the whole vault back to a commit — the recovery case.
func (s *server) rollbackHistory(w http.ResponseWriter, r *http.Request) {
	if !s.history.Available() {
		writeErr(w, http.StatusBadRequest, history.ErrUnavailable.Error())
		return
	}
	var body struct {
		Ref string `json:"ref"`
	}
	if err := decodeBody(r, &body); err != nil {
		handleErr(w, err)
		return
	}
	if strings.TrimSpace(body.Ref) == "" {
		handleErr(w, &vault.ValidationError{Msg: "Pick a version to roll back to."})
		return
	}
	if _, err := s.history.Rollback(body.Ref); err != nil {
		handleHistoryErr(w, err)
		return
	}
	// A rollback rewrites the vault wholesale, so the sync bookkeeping now
	// describes files that are no longer there. Clearing it makes the next run
	// work the difference out from scratch instead of trusting a stale base.
	if s.cloud != nil {
		if err := s.cloud.ForgetSyncState(); err != nil {
			handleErr(w, err)
			return
		}
	}
	s.writeHistory(w, r)
}

// writeHistory answers a write with the log as it now stands, so the client
// never has to guess how its change landed.
func (s *server) writeHistory(w http.ResponseWriter, r *http.Request) {
	commits, err := s.history.Log(historyLimit(r))
	if err != nil {
		handleHistoryErr(w, err)
		return
	}
	writeJSON(w, http.StatusOK, historyStatus{Available: 1, Commits: commits})
}
